test/e2e/crd_e2e_test.go

```diff
@@ -173,15 +170,153 @@ var _ = Describe("CRD Versions", func() {
+		defer cleanupMainCatalogSource()
+
+		// Attempt to get the catalog source before creating install plan
+		_, err := fetchCatalogSourceOnStatus(crc, mainCatalogName, testNamespace, catalogSourceRegistryPodSynced)
+		Expect(err).ToNot(HaveOccurred())
+
+		subscriptionName := genName("sub-nginx-update2-")
+		subscriptionCleanup := createSubscriptionForCatalog(crc, testNamespace, subscriptionName, mainCatalogName, mainPackageName, stableChannel, "", operatorsv1alpha1.ApprovalAutomatic)
+		defer subscriptionCleanup()
+
+		subscription, err := fetchSubscription(crc, testNamespace, subscriptionName, subscriptionHasInstallPlanChecker)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(subscription).ToNot(Equal(nil))
+		Expect(subscription.Status.InstallPlanRef).ToNot(Equal(nil))
+		Expect(oldCSV.GetName()).To(Equal(subscription.Status.CurrentCSV))
+
+		installPlanName := subscription.Status.InstallPlanRef.Name
+
+		// Wait for InstallPlan to be status: Complete before checking resource presence
+		fetchedInstallPlan, err := fetchInstallPlan(GinkgoT(), crc, installPlanName, buildInstallPlanPhaseCheckFunc(operatorsv1alpha1.InstallPlanPhaseComplete))
+		Expect(err).ToNot(HaveOccurred())
+		GinkgoT().Logf("Install plan %s fetched with status %s", fetchedInstallPlan.GetName(), fetchedInstallPlan.Status.Phase)
+		Expect(fetchedInstallPlan.Status.Phase).To(Equal(operatorsv1alpha1.InstallPlanPhaseComplete))
+
+		// old CRD has been installed onto the cluster - now upgrade the subscription to point to the channel with the new CRD
+		// installing the new CSV should fail with a warning about data loss, since a storage version is missing in the new CRD
+		sub, err := crc.OperatorsV1alpha1().Subscriptions(testNamespace).Get(context.TODO(), subscription.GetName(), metav1.GetOptions{})
+		Expect(err).ToNot(HaveOccurred(), "could not get subscription")
+
+		// update channel on sub
+		sub.Spec.Channel = alphaChannel
+		_, err = crc.OperatorsV1alpha1().Subscriptions(testNamespace).Update(context.TODO(), sub, metav1.UpdateOptions{})
+		Expect(err).ToNot(HaveOccurred(), "could not update subscription")
+
+		// fetch new subscription
+		s, err := fetchSubscription(crc, testNamespace, subscriptionName, subscriptionStateAtLatestChecker)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(s).ToNot(Equal(nil))
+		Expect(s.Status.InstallPlanRef).ToNot(Equal(nil))
+
+		// Check the error on the installplan - should be related to data loss and the CRD upgrade missing a stored version
+		Eventually(func() bool {
+			ip, err := crc.OperatorsV1alpha1().InstallPlans(testNamespace).Get(context.TODO(), s.Status.InstallPlanRef.Name, metav1.GetOptions{})
+			Expect(err).ToNot(HaveOccurred(), "could not get installplan")
+
+			Expect(ip.Status.Phase).To(Equal(operatorsv1alpha1.InstallPlanPhaseFailed))
+			Expect(ip.Status.Conditions[len(ip.Status.Conditions)-1].Message).To(ContainSubstring("potential for data loss"))
+			return true
+		}).Should(BeTrue())
```

Comment:
Can we test the case where the user has resolved the issue as well?

i.e. remove the version from old.status.storedVersions and ensure we can update?

Reply:
Sure -- I think I can add another test to verify this behavior.